src/Compilers/Core/MSBuildTask/CheckBootstrapTask.cs

```diff
@@ -0,0 +1,54 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.Build.Utilities;
+using System.IO;
+using System.Reflection;
+using Microsoft.CodeAnalysis.CSharp;
+
+namespace Microsoft.CodeAnalysis.BuildTasks
+{
+#if DEBUG
+    /// <summary>
+    /// This task exists to help us validate our bootstrap building phase is executing correctly.  It 
+    /// is very easy for us, or MSBuild, to accidentally load DLLs from locations that we are 
+    /// not expecting.  This task takes steps to validate the bootstrap phase is executing as expected.
+    /// </summary>
+    public sealed class CheckBootstrapTask : Task
+    {
+        private readonly PropertyInfo _assemblyLocation;
+
+        public CheckBootstrapTask()
+        {
+            _assemblyLocation = typeof(Assembly).GetTypeInfo().GetDeclaredProperty("Location");
+        }
+
+        public override bool Execute()
+        {
+            var thisAssembly = typeof(CheckBootstrapTask).GetTypeInfo().Assembly;
```

Comment:
Something to consider: if you're willing to use assembly version, we could probably use less reflection here and instead look at the version.

I'm not sure if that's more or less robust or the version is better or worse than location, though.